### ISSUE TITLE
Has any one check the executorch implementation on arm (arm_executor_runner.cpp) with mobilenetv2 and real image ?

### ISSUE THREAD

**Issue by dinusha94** [pytorch/executorch#7568]
Hi all,

According to the arm_executor_runner.cpp, it uses a buffer filled with 1. However, when checked with a real image, I get inaccurate classifications. For two separate images, it gave similar classification results. 

Another question I have is, when I check the scaler type for the mv2 model I get it as `Float`, why is that since we use quantization?  As  per my knowledge, to work with Ethos U55 NPU we need an int8 quantized model, please correct me if I am wrong

Thanks

**Comment by mergennachin** [comment]
cc @freddan80 , @zingo, @robell

**Comment by zingo** [comment]
Hi! 
The current example uses float for input as the pytorch model expect it and as it is a somewhat easy/generig/example that can use any model and don't check the output. Hopefully this will improve over time :) 
There are ways to avoid it so you can use 8bit input and skip the inital quantization pass, Ill see if I can dig up some info. 
Also have you a trained the qunatazed step of model? The run.sh flow is kind of bad at it so if you just use it out of the box it probably create very strange/bad quantization parameters.

**Comment by dinusha94** [comment]
Thanks for the clarifications @zingo and @Erik-Lundell , Currently I use simmiler way used in `ml-embedded-evaluation-kit`, where image is hosted as a hex buffer in a .c file, and convert it to `float` buffer before loading to the `input_buffers`  in `arm_executor_runner.cpp` .  following is the function.

```
void setup_input_buffers(std::vector<std::pair<char*, size_t>>& input_buffers) {
    // Fetch image data and metadata
    const uint8_t* image_data = get_sample_data_ptr(0); // Get the data pointer for the image at index 0
    uint32_t img_width = get_sample_img_width(); // Image width
    uint32_t img_height = get_sample_img_height(); // Image height
    uint32_t img_channels = 3; // 3 channels for RGB
    uint32_t total_pixels = img_width * img_height * img_channels;

    if (image_data == nullptr) {
        printf("Failed to retrieve image data from sample files.");
    }

    // Convert image data from uint8_t to float and copy to the new buffer
    size_t idx = 0;
    for (uint32_t i = 0; i < img_height; i++) {
        for (uint32_t j = 0; j < img_width; j++) {
            for (uint32_t c = 0; c < img_channels; c++) {
                float_data[idx++] = static_cast<float>(image_data[(i * img_width + j) * img_channels + c]); // Normalize to [0, 1]
            }
        }
    }

    // Prepare the input buffer entry
    input_buffers.emplace_back(reinterpret_cast<char*>(float_data), total_pixels * sizeof(float));
}
```

As I see in SEMIHOSTING we can provide input as a bin file, `app -m model.pte -i input.bin [-i input2.bin] -o output_basename` ,   

Since you have done validation, can I know that, when creating the bin file can I just use typical image to BIN conversion as following or are there any specific conventions needs to done beforehand.

```
from PIL import Image

def image_to_bin(image_path, output_path):
    # Open the image
    image = Image.open(image_path)
    # Convert the image to grayscale or another desired mode (e.g., "RGB")
    image = image.convert("L")  # "L" for grayscale, "RGB" for color
    # Get raw pixel data
    pixel_data = image.tobytes()
    # Save as a BIN file
    with open(output_path, "wb") as bin_file:
        bin_file.write(pixel_data)
```

**Comment by gggekov** [comment]
@dinusha94 thanks for getting in touch!
What is the datatype of your sensor data? In the ml-embedded-evaluation-kit, at the runtime level the images are in uint8 and are converted to int8 by subtracting 128. If your images are already in (u)int8 like in the MLEK, you can generate a quantised variant of mobilenet_v2 with int8 inputs and populate your input data straight away without converting to float. To do that, you have to modify slightly the `examples/arm/aot_arm_compiler.py` , applying the ioquantization pass [here](https://github.com/pytorch/executorch/blob/main/examples/arm/aot_arm_compiler.py#L563). The modification you need to do is: 
`
from executorch.exir.passes.quantize_io_pass import QuantizeInputs
from executorch.exir.passes.quantize_io_pass import QuantizeOutputs
.....
.....
edge.transform(passes=[QuantizeInputs(edge, [0]),
                      QuantizeOutputs(edge, [0])])
exec_prog = edge.to_executorch(
            config=ExecutorchBackendConfig(
                extract_delegate_segments=False)
        )
print("exported_program = ",exec_prog.exported_program())
`

PS: Side note, but if your images are in (u)int8 in the embedded code, you need to apply the quant scale & zero point to pass to the FP32 domain which I don't think you are doing currently. I recommend you to generate a variant of mobilenet_v2 with int8 inputs straight away.

**Comment by dinusha94** [comment]
@gggekov Thank you for the information, curently my image data is array of uint8_t, array size is (3x224x224) . I will try your instructions and let you know, my end goal is to use this with a camera module in my  board ( Arm® Cortex®-M55 Core with 2× Arm Ethos™-U55 Neural Processing Units— 1× 256 and 1x128 )

**Comment by dinusha94** [comment]
@Erik-Lundell when I try to build `arm_executor_runner` with SEMIHOSTING, it seems like getting stuck and after long time simulation is stoped, To enable SEMIHOSTING 

I make `option(SEMIHOSTING "Enable semihosting" ON)` on CMakeList.txt before running `./run.sh`, 

```
[100%] Built target arm_executor_runner
[build_executorch_runner] Generated baremetal elf file:
/home/dinusha/executorch/examples/arm/executor_runner/cmake-out/arm_executor_runner
executable_text: 203592 bytes
executable_data: 126922636 bytes
executable_bss:  33540 bytes
Running /home/dinusha/executorch/examples/arm/executor_runner/cmake-out/arm_executor_runner for ethos-u55-128 run with FVP:FVP_Corstone_SSE-300_Ethos-U55 num_macs:128
telnetterminal0: Listening for serial connection on port 5000
telnetterminal1: Listening for serial connection on port 5001
telnetterminal2: Listening for serial connection on port 5002
telnetterminal5: Listening for serial connection on port 5003

    Ethos-U rev 136b7d75 --- Apr 12 2023 13:44:01
    (C) COPYRIGHT 2019-2023 Arm Limited
    ALL RIGHTS RESERVED


Info: Simulation is stopping. Reason: CPU time has been exceeded.

Info: /OSCI/SystemC: Simulation stopped by user.
[warning ][main@0][1300 ns] Simulation stopped by user
[run_fvp] Simulation complete, 0
(executorch) dinusha@ubuntu:~/executorch/examples/arm$ 
```

Also it is somewhat unclear how to provide  `model.pte` and `input.bin` to the FVP simulation environment as mentioned in the `arm_executor_runner.cpp`  i.e `app -m model.pte -i input.bin [-i input2.bin] -o output_basename` , current command only accept built binary `arm_executor_runner`

```
${fvp_model}                                            \
            -C ethosu.num_macs=${num_macs}                      \
            -C mps3_board.visualisation.disable-visualisation=1 \
            -C mps3_board.telnetterminal0.start_telnet=0        \
            -C mps3_board.uart0.out_file='-'                    \
            -C mps3_board.uart0.shutdown_on_eot=1               \
            -a "${elf}"                                         \
            --timelimit 220 || true # seconds
```

It would be great if you can provide some example command which you used, I couldn't find ant guide on building and using SEMIHOSTING

**Comment by gggekov** [comment]
Hello @dinusha94,
Apologies for the radio silence over the last few weeks. 

To port the image classification example application from MLEK to ExecuTorch, you need to do a few simple steps. 
1. In MLEK,  the [gen_rgb_cpp.py script](https://review.mlplatform.org/plugins/gitiles/ml/ethos-u/ml-embedded-evaluation-kit/+/refs/heads/main/scripts/py/gen_rgb_cpp.py) allows you to convert an image to a C array of uint8 values. You can reuse this script to generate you C array for the input to the ExecuTorch runtime, but there is one subtlety - by default, the PIL library stores images in channels-last format(Height,Width,Channels). In ExecuTorch, tensors are assumed to be stored in channels-first format(channels, Height, Width) so you need to modify the gen_rgb_cpp.py script to store the C array in channels-first format. You can do that transposing the data in the resize_crop_data.py function [here](https://review.mlplatform.org/plugins/gitiles/ml/ethos-u/ml-embedded-evaluation-kit/+/refs/heads/main/scripts/py/gen_rgb_cpp.py#154).
```
...
np_image = np.array(resized_image, dtype=np.uint8)
np_image_transposed = np.transpose(np_image, (2, 0, 1))   
return np_image_transposed.flatten()
```
Just make the modification above in the python script and let the build system of MLEK invoke gen_rgb_cpp.py for you by doing `cmake .. -DUSE_CASE_BUILD=img_class`. This will generate you C files for each image, for ex. the array for the tiger image is located in `generated/img_class/samples/tiger.c` . 

2. Copy paste the `generated/img_class/samples/tiger.c` into ExecuTorch under `examples/arm/executor_runner/`. Save the array for ex. as `input_data.h` and remove the `#include "BufAttributes.hpp"` and `IFM_BUF_ATTRIBUTE`(these are MLEK-specific libs/defines). I assume the array is called im3. Note that the array size is 150528 because the image resolution is 224x224 in 3 channels(224x224x3 = 150528) 

3. In ExecuTorch, change the `examples/arm/aot_arm_compiler.py` to use the `quantize_io` pass as explained above
```
from executorch.exir.passes.quantize_io_pass import QuantizeInputs
from executorch.exir.passes.quantize_io_pass import QuantizeOutputs
.....
.....
# Apply the QuantizeInputs & QuantizeOutputs passes to input & output tensor 0 
edge.transform(passes=[QuantizeInputs(edge, [0]),
                      QuantizeOutputs(edge, [0])])
exec_prog = edge.to_executorch(
            config=ExecutorchBackendConfig(
                extract_delegate_segments=False)
        )
print("exported_program = ",exec_prog.exported_program())
```
This will generate you a pte file where the input to the model is int8 rather than FP32. We want to do that because the images are stored in uint8 format and we want to avoid requantising them from uint8 to FP32 and then quantise again to int8. Instead, we will convert directly the uint8 images to int8 and feed them to the NN. 

4. In `executorch/examples/arm/executor_runner/arm_executor_runner.cpp` include the .h file from step2. Also, instead of populating 1s to the input buffer, populate the image you saved in step 2. Change [here](https://github.com/pytorch/executorch/blob/main/examples/arm/executor_runner/arm_executor_runner.cpp#L299) to 
````
....
           case ScalarType::Char:
            t.mutable_data_ptr<int8_t>()[j] = im3[j] - 128; //assuming the array with the tiger image is called im3 in the .h 
            break;
....
````
Note that we subtract 128 from the uint8 values for the conversion to int8. 

You are done :) Now run mobilenet_v2(for ex. with `./examples/arm/run.sh --target=ethos-u85-128 --model_name=mv2` for U85) and you will feed the image of the tiger and get classification result from the model. For easier visualisation, I store the result in a set of pair of <int8_t,size_t> and the data is sorted by the int8 value, in this way I can easily see the label with the highest confidence. I am getting highest result on labels 292(corresponding to lion or Panthera leo from the ImageNet dataset) and 282(tabby from ImageNet). 

Two caveats about the differences with MLEK:

- In MLEK, the mobilenet_v2 model we use ends with a softmax whereas the PyTorch model here ends with a linear layer. If you want to convert the NN output into percentage of confidence from 0 to 100%, you would need to manually apply softmax after the PyTorch model. 
- When running mobilenet_v2 via `./examples/arm/run.sh --target=ethos-u85-128 --model_name=mv2`, note that we calibrate the model on a single image(see [here](https://github.com/pytorch/executorch/blob/main/examples/arm/aot_arm_compiler.py#L226)). For more accurate quantisation and more accurate result at runtime, you can calibrate with more images during the quantisation.

**Comment by gggekov** [comment]
Hello @dinusha94, 
Were you able to follow the steps above? You should be able to feed images and get correct classification result.
